eng/packages/http-client-csharp-mgmt/generator/Azure.Generator.Management/src/Utilities/ContextualParameterBuilder.cs

```diff
@@ -0,0 +1,147 @@
+                            else
+                            {
+                                // we have to reassign the value of parentLayerCount to a local variable to avoid the closure to wrap the parentLayerCount variable which changes during recursion.
+                                int currentParentCount = parentLayerCount;
+                                parameterStack.Push(new ContextualParameter(key.Value, value.VariableName, id => BuildParentInvocation(currentParentCount, id).ResourceType().Namespace()));
+                            }
+                            // do not append a new .Parent to the id
+                        }
+                        else // for all other normal keys
+                        {
+                            // we have to reassign the value of parentLayerCount to a local variable to avoid the closure to wrap the parentLayerCount variable which changes during recursion.
+                            int currentParentCount = parentLayerCount;
+                            parameterStack.Push(new ContextualParameter(key.Value, value.VariableName, id => BuildParentInvocation(currentParentCount, id).Name()));
+                            appendParent = true;
+                        }
+                    }
+                    else // in this branch value is a constant
+                    {
+                        if (key.IsProvidersSegment)
+                        {
+                            // if the key is not providers, we need to skip this level and increment the parent hierarchy
+                            appendParent = true;
+                        }
+                    }
+                }
+                // check if we need to call .Parent on id
+                if (appendParent)
+                {
+                    parentLayerCount++;
+                }
+            }
+            // recursively get the parameters of its parent
+            BuildContextualParameterHierarchy(parent, parameterStack, parentLayerCount);
+
+            static ScopedApi<ResourceIdentifier> BuildParentInvocation(int parentLayerCount, ScopedApi<ResourceIdentifier> id)
+            {
+                var result = id;
+                for (int i = 0; i < parentLayerCount; i++)
+                {
+                    result = result.Parent();
+                }
+                return result;
+            }
+        }
+
+        private static IReadOnlyList<KeyValuePair<RequestPathSegment, RequestPathSegment>> SplitIntoPairs(IReadOnlyList<RequestPathSegment> requestPath)
+        {
+            // in our current cases, we will always have even segments.
+            Debug.Assert(requestPath.Count % 2 == 0, "The request path should always have an even number of segments for pairing.");
+            int maxNumberOfPairs = requestPath.Count / 2;
+            var pairs = new KeyValuePair<RequestPathSegment, RequestPathSegment>[maxNumberOfPairs];
+
+            for (int i = 0; i < maxNumberOfPairs; i++)
```

Comment:
can we start from maxNumberOfPairs - 1, so that we can get rid of Reverse() in L69